File: 2024-4-16/Add One Row to Tree.js
var addOneRow = function(root, val, depth) {
    const helper = (node, val, depth, currdepth) => {
        if (depth === 1) {
            const newRoot = new TreeNode(val);
            newRoot.left = node;
            return newRoot;
        }

        if (!node) {
            return null;
        }

        if (currdepth === depth - 1) {
            const leftman = node.left;
            const rightman = node.right;

            node.left = new TreeNode(val);
            node.left.left = leftman;
            node.left.right = null;

            node.right = new TreeNode(val);
            node.right.left = null;
            node.right.right = rightman;
            
            return node;
        }

        node.left = helper(node.left, val, depth, currdepth + 1);
        node.right = helper(node.right, val, depth, currdepth + 1);

        return node;
    };

    return helper(root, val, depth, 1);
};

File: 2024-4-16/Add One Row to Tree.py
class Solution:
    def helper(self, root: TreeNode, val: int, depth: int, currdepth: int) -> TreeNode:
        if depth == 1:
            new_root = TreeNode(val)
            new_root.left = root
            return new_root

        if not root:
            return root

        if currdepth == depth - 1:
            leftman = root.left
            rightman = root.right

            root.left = TreeNode(val)
            root.right = TreeNode(val)

            root.left.left = leftman
            root.right.right = rightman
            return root

        root.left = self.helper(root.left, val, depth, currdepth + 1)
        root.right = self.helper(root.right, val, depth, currdepth + 1)

        return root

    def addOneRow(self, root: TreeNode, val: int, depth: int) -> TreeNode:
        return self.helper(root, val, depth, 1)